dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/CommittingPartitionTracker.java

```diff
@@ -0,0 +1,304 @@
+            if (Thread.interrupted()) {
+                // Don't do anything if we were cancelled
+                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                    Tr.debug(this, tc, "Commit task running but has been cancelled", this);
+                }
+                return;
+            }
+
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(this, tc, "Checking for new work to commit, last committed offset is " + committedOffset, this);
+                Tr.debug(this, tc, "Current completed work", completedWork);
+            }
+
+            long newCommitOffset = committedOffset;
+            CompletedWork newestWork = null;
+            for (CompletedWork work : completedWork) {
+                if (work.offset < newCommitOffset) {
+                    // Work that we've already asked to commit, ignore it
+                } else if (work.offset == newCommitOffset) {
+                    // Work that should be committed now
+                    newCommitOffset++;
+                    newestWork = work;
+                } else {
+                    // We've reached the end of a continuous block of completed work
+                    // Can't commit any further work until everything before it has been committed
+                    break;
+                }
+            }
+
+            if (newestWork != null) {
+                // commit
+                long originalCommitOffset = committedOffset;
+                long finalCommitOffset = newCommitOffset;
+
+                if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                    Tr.event(this, tc, "Committing from " + originalCommitOffset + " to " + finalCommitOffset, this);
+                }
+                commitUpTo(newestWork).whenCompleteAsync((r, t) -> processCommittedWork(originalCommitOffset, finalCommitOffset, t), executor);
+            }
+
+            outstandingUncommittedWork -= newCommitOffset - committedOffset;
+            pendingCommitTask = null;
+            committedOffset = newCommitOffset;
+        }
+    }
+
+    /**
+     * Commit the offset up to the offset of the given work
+     *
+     * @param work
+     * @return completion stage which completes with the result of the commit, or completes exceptionally if the commit failed
+     */
+    private CompletionStage<Void> commitUpTo(CompletedWork work) {
+        if (isClosed()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(this, tc, "Rejecting commit attempt because partition is closed", this);
+            }
+
+            CompletableFuture<Void> result = new CompletableFuture<>();
+            result.completeExceptionally(new Exception("Partition is closed"));
+            return result;
+        }
+
+        // Note work.offset + 1
+        // Committed offset is the _next_ message we expect to receive
```

Comment:
huh? `Committed` is past tense, `next message` is future??

Reply:
Does this make more sense:
```java
        // Note work.offset + 1
        // In general the committed offset for a partition is the first message which should be received by a new consumer
        // which starts consuming from that partition.
        // Therefore, the offset which we are about to commit must be the offset _after_ the last message we have processed.
```

Reply:
If the committed offset is the first message which should be received by a new consumer ... does that mean it's counting from 1? There is no message "index" 0?

Reply:
No, the offset of the first message is `0`.

Imagine there are three messages in the partition, none have yet been consumed. The committed offset is `0`, the next message to be received has offset `0`.

The first message is read and processed, the consumer doesn't want to receive this message again. The consumer has processed the record with offset `0`, it sets the committed offset to `1`, the next message to be received has offset `1`.